File: source/lexer.cxx
#include "token.hxx"

char Lexer::peek(size_t offset) const
{
    size_t p = Position + offset;
    return p < Source.size() ? Source[p] : '\0';
}

char Lexer::advance()
{
    if (Position >= Source.size())
        return '\0';

    char c = Source[Position++];
    if (c == '\n')
    {
        Line++;
        Column = 1;
    }
    else
        Column++;
    return c;
}

void Lexer::skipWhitespace()
{
    while (std::isspace(peek()))
        advance();
}

Token Lexer::makeToken(TokenType type, size_t start, size_t end, size_t line, size_t column) const
{
    return Token{type, std::string_view(Source.data() + start, end - start), File, line, column};
}

TokenType Lexer::lookupIdentifier(std::string_view s) const
{
    if (auto it = keywords.find(s); it != keywords.end())
    {
        return it->second;
    }
    return TokenType::Identifier;
}

Token Lexer::scanIdentifier(size_t line, size_t column)
{
    size_t start = Position;
    while (std::isalnum(peek()) || peek() == '_' || peek() == '\'')
        advance();
    size_t end = Position;
    std::string_view ident(Source.data() + start, end - start);
    TokenType type = lookupIdentifier(ident);
    return makeToken(type, start, end, line, column);
}

Token Lexer::scanString(size_t line, size_t column)
{
    size_t start = Position;
    advance();
    while (peek() != '\0')
    {
        if (peek() == '"')
        {
            advance();
            break;
        }
        if (peek() == '\\')
        {
            advance();
            if (peek() != '\0')
                advance();
            continue;
        }
        advance();
    }
    return makeToken(TokenType::String, start, Position, line, column);
}

Token Lexer::scanByte(size_t line, size_t column)
{
    size_t start = Position;
    advance();
    if (peek() == '\\')
    {
        advance();
        if (peek() != '\0')
            advance();
    }
    else if (peek() != '\0')
        advance();
    if (peek() == '\'')
        advance();
    return makeToken(TokenType::Byte, start, Position, line, column);
}

Token Lexer::scanNumber(size_t line, size_t column)
{
    size_t start = Position;
    bool isFloat = false;
    while (std::isdigit(peek()) || peek() == '.')
    {
        if (peek() == '.')
        {
            if (isFloat)
                break;
            isFloat = true;
        }
        advance();
    }
    if (peek() == 'u')
    {
        advance();
        return makeToken(TokenType::Unsigned, start, Position, line, column);
    }
    return makeToken(isFloat ? TokenType::Float : TokenType::Integer, start, Position, line, column);
}

Token Lexer::scanLineComment(size_t line, size_t column)
{
    size_t start = Position;
    advance();
    advance();
    while (peek() != '\0' && peek() != '\n')
        advance();
    size_t end = Position;
    while (end > start && std::isspace(static_cast<unsigned char>(Source[end - 1])))
        --end;
    return makeToken(TokenType::Comment, start, end, line, column);
}

int Lexer::precedence(TokenType type) const
{
    switch (type)
    {
    case TokenType::Vbar:
    case TokenType::Colon:
        return 1;
    case TokenType::And:
        return 2;
    case TokenType::Equal:
    case TokenType::NotEqual:
        return 3;
    case TokenType::LessThan:
    case TokenType::LessEqual:
    case TokenType::GreaterThan:
    case TokenType::GreaterEqual:
        return 4;
    case TokenType::Plus:
    case TokenType::Minus:
        return 5;
    case TokenType::Asterisk:
    case TokenType::Slash:
    case TokenType::Percent:
        return 6;
    default:
        return 0;
    }
}

Token Lexer::next()
{
    skipWhitespace();
    size_t line = Line;
    size_t column = Column;
    char c = peek();

    if (c == '\0')
        return makeToken(TokenType::EndOfFile, Position, Position, line, column);
    if (std::isalpha(c) || c == '_')
        return scanIdentifier(line, column);
    if (c == '"')
        return scanString(line, column);
    if (c == '\'')
        return scanByte(line, column);
    if (std::isdigit(c))
        return scanNumber(line, column);
    if (c == '/' && peek(1) == '/')
        return scanLineComment(line, column);

    advance();

    switch (c)
    {
    case '=':
        return peek() == '=' ? (advance(), makeToken(TokenType::Equal, Position - 2, Position, line, column))
                             : makeToken(TokenType::Assign, Position - 1, Position, line, column);
    case '!':
        return peek() == '=' ? (advance(), makeToken(TokenType::NotEqual, Position - 2, Position, line, column))
                             : makeToken(TokenType::Not, Position - 1, Position, line, column);
    case '<':
        return peek() == '=' ? (advance(), makeToken(TokenType::LessEqual, Position - 2, Position, line, column))
                             : makeToken(TokenType::LessThan, Position - 1, Position, line, column);
    case '>':
        return peek() == '=' ? (advance(), makeToken(TokenType::GreaterEqual, Position - 2, Position, line, column))
                             : makeToken(TokenType::GreaterThan, Position - 1, Position, line, column);
    case '&':
        return peek() == '&' ? (advance(), makeToken(TokenType::And, Position - 2, Position, line, column))
                             : makeToken(TokenType::Illegal, Position - 1, Position, line, column);
    case '|':
        return peek() == '|' ? (advance(), makeToken(TokenType::Or, Position - 2, Position, line, column))
                             : makeToken(TokenType::Vbar, Position - 1, Position, line, column);
    case '+':
        return makeToken(TokenType::Plus, Position - 1, Position, line, column);
    case '-':
        return makeToken(TokenType::Minus, Position - 1, Position, line, column);
    case '*':
        return makeToken(TokenType::Asterisk, Position - 1, Position, line, column);
    case '/':
        return makeToken(TokenType::Slash, Position - 1, Position, line, column);
    case '%':
        return makeToken(TokenType::Percent, Position - 1, Position, line, column);
    case '(':
        return makeToken(TokenType::LeftParen, Position - 1, Position, line, column);
    case ')':
        return makeToken(TokenType::RightParen, Position - 1, Position, line, column);
    case '{':
        return makeToken(TokenType::LeftBrace, Position - 1, Position, line, column);
    case '}':
        return makeToken(TokenType::RightBrace, Position - 1, Position, line, column);
    case '[':
        return makeToken(TokenType::LeftBracket, Position - 1, Position, line, column);
    case ']':
        return makeToken(TokenType::RightBracket, Position - 1, Position, line, column);
    case ',':
        return makeToken(TokenType::Comma, Position - 1, Position, line, column);
    case ';':
        return makeToken(TokenType::Semicolon, Position - 1, Position, line, column);
    case ':':
        return makeToken(TokenType::Colon, Position - 1, Position, line, column);
    case '.':
        return makeToken(TokenType::Dot, Position - 1, Position, line, column);
    default:
        return makeToken(TokenType::Illegal, Position - 1, Position, line, column);
    }
}

File: internal/lexer/lexer.go
package lexer

import (
	"fmt"
	"strconv"
	"unicode"
)

type Lexer struct {
	source       []rune
	position     int
	lineNumber   int
	columnNumber int
	fileName     string
}

func Tokenize(source, filename string) ([]Token, error) {
	lexer := New(source, filename)
	out := []Token{}
	for {
		token := lexer.Next()
		out = append(out, token)
		if token.Type == EndOfFile {
			break
		}
		if token.Type == Illegal {
			lexer.error("Illegal character")
		}
	}
	return out, nil
}

func New(source, filename string) *Lexer {
	return &Lexer{
		source:       []rune(source),
		position:     0,
		lineNumber:   1,
		columnNumber: 1,
		fileName:     filename,
	}
}

func (lexer *Lexer) Next() Token {
	lexer.consumeWhitespace()
	startLine, startColumn := lexer.lineNumber, lexer.columnNumber
	char := lexer.peekRuneAt(0)

	if char == 0 {
		return lexer.makeToken(EndOfFile, "", startLine, startColumn)
	}

	if isIdentifierStart(char) {
		lex := lexer.scanIdentifier()
		type_ := LookupIdentifier(lex)
		return lexer.makeToken(type_, lex, startLine, startColumn)
	}

	if unicode.IsDigit(char) {
		lex, type_ := lexer.scanNumberLiteral()
		return lexer.makeToken(type_, lex, startLine, startColumn)
	}

	if char == '"' {
		lex := lexer.scanStringLiteral()
		return lexer.makeToken(String, lex, startLine, startColumn)
	}

	if char == '\'' {
		lex := lexer.scanByteLiteral()
		return lexer.makeToken(Byte, lex, startLine, startColumn)
	}

	if char == '/' && lexer.peekRuneAt(1) == '/' {
		lex := lexer.scanLineComment()
		return lexer.makeToken(Comment, lex, startLine, startColumn)
	}

	switch char {
	case '=':
		lexer.advanceRune()
		if lexer.peekRuneAt(0) == '=' {
			lexer.advanceRune()
			return lexer.makeToken(Equal, "==", startLine, startColumn)
		}
		return lexer.makeToken(Assign, "=", startLine, startColumn)
	case '!':
		lexer.advanceRune()
		if lexer.peekRuneAt(0) == '=' {
			lexer.advanceRune()
			return lexer.makeToken(NotEqual, "!=", startLine, startColumn)
		}
		return lexer.makeToken(Not, "!", startLine, startColumn)
	case '<':
		lexer.advanceRune()
		if lexer.peekRuneAt(0) == '=' {
			lexer.advanceRune()
			return lexer.makeToken(LessEqual, "<=", startLine, startColumn)
		}
		return lexer.makeToken(LessThan, "<", startLine, startColumn)
	case '>':
		lexer.advanceRune()
		if lexer.peekRuneAt(0) == '=' {
			lexer.advanceRune()
			return lexer.makeToken(GreaterEqual, ">=", startLine, startColumn)
		}
		return lexer.makeToken(GreaterThan, ">", startLine, startColumn)
	case '+':
		lexer.advanceRune()
		return lexer.makeToken(Plus, "+", startLine, startColumn)
	case '-':
		lexer.advanceRune()
		return lexer.makeToken(Minus, "-", startLine, startColumn)
	case '*':
		lexer.advanceRune()
		return lexer.makeToken(Asterisk, "*", startLine, startColumn)
	case '/':
		lexer.advanceRune()
		return lexer.makeToken(Slash, "/", startLine, startColumn)
	case '%':
		lexer.advanceRune()
		return lexer.makeToken(Percent, "%", startLine, startColumn)
	case ':':
		lexer.advanceRune()
		return lexer.makeToken(Colon, ":", startLine, startColumn)
	case ',':
		lexer.advanceRune()
		return lexer.makeToken(Comma, ",", startLine, startColumn)
	case '{':
		lexer.advanceRune()
		return lexer.makeToken(LeftBrace, "{", startLine, startColumn)
	case '}':
		lexer.advanceRune()
		return lexer.makeToken(RightBrace, "}", startLine, startColumn)
	case '(':
		lexer.advanceRune()
		return lexer.makeToken(LeftParen, "(", startLine, startColumn)
	case ')':
		lexer.advanceRune()
		return lexer.makeToken(RightParen, ")", startLine, startColumn)
	case '[':
		lexer.advanceRune()
		return lexer.makeToken(LeftBracket, "[", startLine, startColumn)
	case ']':
		lexer.advanceRune()
		return lexer.makeToken(RightBracket, "]", startLine, startColumn)
	case '.':
		lexer.advanceRune()
		return lexer.makeToken(Dot, ".", startLine, startColumn)
	case ';':
		lexer.advanceRune()
		return lexer.makeToken(Semicolon, ";", startLine, startColumn)
	case '&':
		lexer.advanceRune()
		if lexer.peekRuneAt(0) == '&' {
			lexer.advanceRune()
			return lexer.makeToken(And, "&&", startLine, startColumn)
		}
		return lexer.makeToken(Illegal, "&", startLine, startColumn)
	case '|':
		lexer.advanceRune()
		if lexer.peekRuneAt(0) == '|' {
			lexer.advanceRune()
			return lexer.makeToken(Or, "||", startLine, startColumn)
		}
		return lexer.makeToken(Vbar, "|", startLine, startColumn)
	default:
		r := lexer.advanceRune()
		return lexer.makeToken(Illegal, string(r), startLine, startColumn)
	}
}

func (lexer *Lexer) advanceRune() rune {
	if lexer.position >= len(lexer.source) {
		return 0
	}
	r := lexer.source[lexer.position]
	lexer.position++
	if r == '\n' {
		lexer.lineNumber++
		lexer.columnNumber = 1
	} else {
		lexer.columnNumber++
	}
	return r
}

func (lexer *Lexer) peekRuneAt(offset int) rune {
	index := lexer.position + offset
	if index >= len(lexer.source) {
		return 0
	}
	return lexer.source[index]
}

func (lexer *Lexer) makeToken(_type TokenType, lexeme string, line, column int) Token {
	return Token{
		Type:   _type,
		Lexeme: lexeme,
		Line:   line,
		Column: column,
		File:   lexer.fileName,
		Source: lexer.source,
	}
}

func (lexer *Lexer) scanIdentifier() string {
	start := lexer.position
	for {
		ch := lexer.peekRuneAt(0)
		if ch == 0 || !(isIdentifierPart(ch)) {
			break
		}
		lexer.advanceRune()
	}
	return string(lexer.source[start:lexer.position])
}

func (lexer *Lexer) scanNumberLiteral() (string, TokenType) {
	start := lexer.position
	isFloat := false
	for {
		ch := lexer.peekRuneAt(0)
		if ch == '.' {
			if lexer.peekRuneAt(1) == '.' {
				break
			}
			isFloat = true
			lexer.advanceRune()
			continue
		}
		if !unicode.IsDigit(ch) && ch != '_' {
			break
		}
		lexer.advanceRune()
	}
	lex := string(lexer.source[start:lexer.position])
	clean := StripNumericSeparators(lex)
	if isFloat {
		if _, err := strconv.ParseFloat(clean, 64); err == nil {
			return lex, Float
		}
	}
	if _, err := strconv.ParseInt(clean, 10, 64); err == nil {
		return lex, Integer
	}
	if _, err := strconv.ParseFloat(clean, 64); err == nil {
		return lex, Float
	}
	return lex, Illegal
}

func (lexer *Lexer) scanStringLiteral() string {
	quote := lexer.advanceRune()
	start := lexer.position - 1
	if quote != '"' {
		return string(lexer.source[start:lexer.position])
	}
	ch := lexer.advanceRune()
	if ch == 0 {
		lexer.error("unterminated string literal")
		return string(lexer.source[start:lexer.position])
	}
	if ch == quote {
		lexer.error("empty string literal")
		return string(lexer.source[start:lexer.position])
	}
	runeCount := 0
	for {
		if ch == '\\' {
			runeCount++
			esc := lexer.advanceRune()
			if esc == 0 {
				lexer.error("unterminated escape sequence in string literal")
				return string(lexer.source[start:lexer.position])
			}
			switch esc {
			case 'n', 't', 'r', '\\', '\'', '"', '0':
			default:
				lexer.error(fmt.Sprintf("invalid escape character: \\%c", esc))
				return string(lexer.source[start:lexer.position])
			}
		} else {
			runeCount++
		}
		ch = lexer.advanceRune()
		if ch == 0 {
			lexer.error("unterminated string literal")
			return string(lexer.source[start:lexer.position])
		}
		if ch == quote {
			break
		}
	}
	return string(lexer.source[start:lexer.position])
}

func (lexer *Lexer) scanByteLiteral() string {
	quote := lexer.advanceRune()
	start := lexer.position - 1
	if quote != '\'' {
		return string(lexer.source[start:lexer.position])
	}
	ch := lexer.advanceRune()
	if ch == 0 {
		lexer.error("unterminated character literal")
		return string(lexer.source[start:lexer.position])
	}
	if ch == quote {
		lexer.error("empty character literal")
		return string(lexer.source[start:lexer.position])
	}
	var runeCount int = 0
	if ch == '\\' {
		runeCount++
		esc := lexer.advanceRune()
		if esc == 0 {
			lexer.error("unterminated escape sequence in character literal")
			return string(lexer.source[start:lexer.position])
		}
		switch esc {
		case 'n', 't', 'r', '\\', '\'', '"', '0':
		default:
			lexer.error(fmt.Sprintf("invalid escape character: \\%c", esc))
			return string(lexer.source[start:lexer.position])
		}
	} else {
		runeCount++
	}
	end := lexer.advanceRune()
	if end == 0 {
		lexer.error("unterminated character literal")
		return string(lexer.source[start:lexer.position])
	}
	if end != '\'' {
		lexer.error("extra characters in character literal (expected closing ')")
		return string(lexer.source[start:lexer.position])
	}
	if runeCount != 1 {
		lexer.error("character literal must contain exactly 1 character or valid escape")
	}
	return string(lexer.source[start:lexer.position])
}

func (lexer *Lexer) scanLineComment() string {
	start := lexer.position
	lexer.advanceRune()
	lexer.advanceRune()
	for {
		ch := lexer.peekRuneAt(0)
		if ch == 0 || ch == '\n' {
			break
		}
		lexer.advanceRune()
	}
	return string(lexer.source[start:lexer.position])
}

func (lexer *Lexer) consumeWhitespace() {
	for {
		ch := lexer.peekRuneAt(0)
		if ch == 0 {
			return
		}
		if ch == ' ' || ch == '\t' || ch == '\r' || ch == '\n' {
			lexer.advanceRune()
			continue
		}
		break
	}
}
